### PR TITLE
For #2648: Updates trimming of url hostname

### PR DESCRIPTION
**Pull request by sblatz**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt [hunk 1]
```diff
@@ -81,31 +82,33 @@ class CollectionViewHolder(
...
-            } else {
-                tabsDisplayed += 1
-                it
+        launch(Dispatchers.Main) {
+            view.collection_item.text = collection.title
+
+            val hostNameList = collection.tabs.map { it.url.urlToTrimmed(view.context).capitalize() }
+
+            var tabsDisplayed = 0
+            val tabTitlesList = hostNameList.joinToString(", ") {
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt]: Nit: eliminate `hostNameList`

```kotlin
val collectionDescription = collection.tabs
   .map { it.url.urlToTrimmed(view.context) }
   .map { if (it.length > etc ) … }
   .map { it.capitalize() }
   .joinToString(", ") 

view.collection_description.text = collectionDescription
```

File: app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt [hunk 2]
```diff
@@ -81,31 +82,33 @@ class CollectionViewHolder(
...
-        view.collection_icon.setColorFilter(
-            ContextCompat.getColor(
-                view.context,
-                getIconColor(collection.id)
-            ),
-            android.graphics.PorterDuff.Mode.SRC_IN
-        )
     }
 
     @Suppress("ComplexMethod", "MagicNumber")
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt]: Hmm, ok. 

Not part of this PR, but I think an array of colors and a modulo index would've been better here, I think.

File: app/src/main/java/org/mozilla/fenix/ext/String.kt [hunk 3]
```diff
@@ -28,21 +31,6 @@ fun String?.getHostFromUrl(): String? = try {
...
-        }
-    } catch (e: MalformedURLException) {
-        this.getHostFromUrl() ?: ""
-    } catch (e: StringIndexOutOfBoundsException) {
-        this.getHostFromUrl() ?: ""
-    }
-}
+suspend fun String.urlToTrimmed(context: Context): String {
+    return PublicSuffixList(context).stripPublicSuffix(URL(this).host).await().removePrefix("www.")
+}
```
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/org/mozilla/fenix/ext/String.kt]: Hmm. Yeah I see why you have this as a string method. I worry about error handling here.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 4]
```diff
@@ -527,70 +529,76 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
...
+                view!!, getString(R.string.snackbar_tab_deleted),
+                getString(R.string.snackbar_deleted_undo), {
+                    deleteSessionJob = null
+                    emitSessionChanges()
                 }
+            ) {
+                sessionManager.findSessionById(sessionId)
+                    ?.let { session ->
+                        sessionManager.remove(session)
+                    }
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: Nit: put these 4 lines of code into a closure `val` and use it here and `deleteSessionJob` above.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt [hunk 5]
```diff
@@ -527,70 +529,76 @@ class HomeFragment : Fragment(), CoroutineScope, AccountObserver {
...
-            )
-        )
+        context?.let { context ->
+            launch(Dispatchers.Main) {
+                val sessionManager = requireComponents.core.sessionManager
+
+                getManagedEmitter<SessionControlChange>().onNext(
+                    SessionControlChange.TabsChange(
+                        sessionManager.sessions
+                            .filter { (activity as HomeActivity).browsingModeManager.isPrivate == it.private }
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: Nit: put 

```kotlin
(activity as HomeActivity).browsingModeManager.isPrivate
```

into a local variable.

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/CollectionViewHolder.kt [hunk 6]
```diff
@@ -92,47 +93,49 @@ class CollectionViewHolder(
...
+            val tabTitlesList = hostNameList.joinToString(", ") {
+                if (it.length > maxTitleLength) {
+                    it.substring(
+                        0,
+                        maxTitleLength
+                    ) + "..."
+                } else {
+                    tabsDisplayed += 1
+                    it
+                }
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/CollectionViewHolder.kt]: Nit: put this in a `fun Collection<Tab>.description()` method.

File: app/src/main/java/org/mozilla/fenix/ext/String.kt [hunk 7]
```diff
@@ -28,21 +30,6 @@ fun String?.getHostFromUrl(): String? = try {
...
-            firstIndex - 1 == lastIndex -> url.host.substring(0, lastIndex)
-            firstIndex < lastIndex -> url.host.substring(firstIndex, lastIndex)
-            else -> url.host
-        }
-    } catch (e: MalformedURLException) {
-        this.getHostFromUrl() ?: ""
-    } catch (e: StringIndexOutOfBoundsException) {
-        this.getHostFromUrl() ?: ""
-    }
+suspend fun String.urlToTrimmedHost(context: Context): String {
```
Comment (sblatz) [inline app/src/main/java/org/mozilla/fenix/ext/String.kt]: Is this an appropriate way to handle this method, @pocmo? I tried to pattern match what you implemented for `URLRenderer` in A-C.

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt (hunk 5 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: What's the benefit in doing this 🤔

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt (hunk 4 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: 👍

File: app/src/main/java/org/mozilla/fenix/ext/String.kt (hunk 7 above)
Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/ext/String.kt]: (1) Turn `PublicSuffixList` into a Singleton (like the other components that you can reach from a `Context` object, e.g. `SessionManager`). It is a pretty heavy thing that needs to load a bunch of things from disk. You don't want to do that multiple times and you definitely don't want to have multiple of those hanging around.

(2) Regarding the prefix: There used to be a method in Fennec, maybe Focus and should be in AC (but I can't find it.. if it doesn't exist we should create it eventually) that removes common prefixes, e.g. not only "www", but also "m" for mobile pages.

(3) Consider the edge cases: Not everything is a http(s) URL. What if the URL has no host? What about IP addresses? I'm not sure whether `PublicSuffixList` can deal with all those cases since it expects you to pass in a host and may not handle wonky values well.

File: app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt (hunk 2 above)
Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt]: This reminds me that we talked about providing this in the AC API, but I actually never implemented that.

Also `id % 4` should never hit the case for `4` and `else` - so I assume `yellow` and `white` are not actually used right now?

File: app/src/main/java/org/mozilla/fenix/ext/String.kt (hunk 7 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/ext/String.kt]: Found the prefix stripping in Focus: https://github.com/mozilla-mobile/focus-android/blob/master/app/src/main/java/org/mozilla/focus/utils/UrlUtils.java#L142

Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/ext/String.kt]: Regarding 3) I tested with IP addresses and PublicSuffixList seems to handle it fine (the whole string is just returned e.g. `192.168.1.254`

File: app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt (hunk 2 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/collections/SaveCollectionListAdapter.kt]: Good catch, @pocmo--sometimes my brain forgets modulo math 🙄. I've updated it to be 5, but yes, white should never be used.

File: app/src/main/java/org/mozilla/fenix/ext/String.kt (hunk 7 above)
Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/ext/String.kt]: A-C issue: https://github.com/mozilla-mobile/android-components/issues/3299

**Comment by sblatz** [comment]
@pocmo @jhugman I believe I've addressed the concerns for this PR 😄

File: app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt (hunk 5 above)
Comment (jhugman) [reply app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt]: It's a nit, so I'm not insanely worried about it. However:

 * The expression can be calculated once, but it's currently being calculated for every session in the `sessionManager`. It's possible that the compiler knows that both the `browsingModeManager` and `isPrivate` properties are both side effect free, so can optimize it for you, but I don't know.

 * From a style point of view: having a cast and then two dots in a small closure makes the line long — and in narrow windows like the diff viewer — slightly harder to read :(

File: app/src/main/java/org/mozilla/fenix/ext/String.kt [hunk 8]
```diff
@@ -28,21 +29,26 @@ fun String?.getHostFromUrl(): String? = try {
...
+ */
+suspend fun String.urlToTrimmedHost(context: Context): String {
+    return try {
+        val host = URL(this).host.stripCommonPrefixes()
+        context.components.publicSuffixList.stripPublicSuffix(host).await()
     } catch (e: MalformedURLException) {
-        this.getHostFromUrl() ?: ""
-    } catch (e: StringIndexOutOfBoundsException) {
-        this.getHostFromUrl() ?: ""
+        this
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/ext/String.kt]: 🎉

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/CollectionViewHolder.kt [hunk 9]
```diff
@@ -92,47 +93,47 @@ class CollectionViewHolder(
...
+                .map {
+                    if (it.length > maxTitleLength) {
+                        it.substring(
+                            0,
+                            maxTitleLength
+                        ) + "…"
+                    } else {
+                        it
+                    }
+                }.joinToString(", ")
```
Comment (jhugman) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/CollectionViewHolder.kt]: This code and [this code][1] is identical.

I'd put this in an override function for 

```kotlin
fun Collection<Tab>.description(): String {

}
```
so you can use it in two places.

Either do it now ✅, or file an issue for it ✅.

[1]: https://github.com/mozilla-mobile/fenix/pull/3191/files#diff-ae7a7f880e17ab61958437bea4fd1e89R88

Comment (sblatz) [reply app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/CollectionViewHolder.kt]: Done! 😄
